ICE: no type for local variable local _=m

I'm seeing an internal compiler error on the following input, found by [fuzz-rustc](https://github.com/dwrensha/fuzz-rustc):

### Code

```Rust
fn r({){Ok{d..||_=m
```

### Error output

```
error: internal compiler error: compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs:139:13: no type for local variable local _=m (hir_id=HirId { owner: DefId(0:3 ~ playground[b6e0]::r), local_id: 6 })
 --> src/lib.rs:1:17
  |
1 | fn r({){Ok{d..||_=m
  |                 ^^^

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/cfa3fe5af339e724209b25715282adae0c61628f/compiler/rustc_errors/src/lib.rs:1115:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The ICE happens after reporting some other errors, and the input is not syntactically correct (`fn r({)`), so low priority. Affected versions: I tried nightly 2021-12-31 and stable 1.57 and they both show the same ICE.

<details><summary><strong>Backtrace</strong></summary>
<p>

```
   Compiling playground v0.0.1 (/playground)
error: this file contains an unclosed delimiter
 --> src/main.rs:1:20
  |
1 | fn r({){Ok{d..||_=m
  |        -  -        ^
  |        |  |
  |        |  unclosed delimiter
  |        unclosed delimiter

error: expected parameter name, found `{`
 --> src/main.rs:1:6
  |
1 | fn r({){Ok{d..||_=m
  |      ^ expected parameter name

error: expected one of `,` or `}`, found `..`
 --> src/main.rs:1:6
  |
1 | fn r({){Ok{d..||_=m
  |      ^      -^
  |      |      |
  |      |      help: `}` may belong here
  |      unclosed delimiter

error[E0425]: cannot find value `d` in this scope
 --> src/main.rs:1:12
  |
1 | fn r({){Ok{d..||_=m
  |            ^ not found in this scope

error[E0425]: cannot find value `m` in this scope
 --> src/main.rs:1:19
  |
1 | fn r({){Ok{d..||_=m
  |                   ^ not found in this scope

error[E0559]: variant `Result<_, _>::Ok` has no field named `d`
   --> src/main.rs:1:12
    |
1   | fn r({){Ok{d..||_=m
    |            ^ field does not exist
    |
help: `Result<_, _>::Ok` is a tuple variant, use the appropriate syntax
    |
1   | fn r({){Result<_, _>::Ok(/* fields */)
    |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0308]: mismatched types
 --> src/main.rs:1:9
  |
1 | fn r({){Ok{d..||_=m
  |        -^^^^^^^^^^^ expected `()`, found enum `Result`
  |        |
  |        possibly return type missing here?
  |
  = note: expected unit type `()`
                  found enum `Result<_, _>`

error: internal compiler error: compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs:139:13: no type for local variable local _=m (hir_id=HirId { owner: DefId(0:3 ~ playground[87de]::r), local_id: 6 })
 --> src/main.rs:1:17
  |
1 | fn r({){Ok{d..||_=m
  |                 ^^^

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/cfa3fe5af339e724209b25715282adae0c61628f/compiler/rustc_errors/src/lib.rs:1115:9
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: std::panic::panic_any::<rustc_errors::ExplicitBug>
   2: <rustc_errors::HandlerInner>::span_bug::<rustc_span::span_encoding::Span>
   3: <rustc_errors::Handler>::span_bug::<rustc_span::span_encoding::Span>
   4: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, ()>
   5: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>
   6: rustc_middle::util::bug::span_bug_fmt::<rustc_span::span_encoding::Span>
   7: <rustc_typeck::check::writeback::WritebackCx as rustc_hir::intravisit::Visitor>::visit_local
   8: <rustc_typeck::check::writeback::WritebackCx as rustc_hir::intravisit::Visitor>::visit_expr
   9: <rustc_typeck::check::writeback::WritebackCx as rustc_hir::intravisit::Visitor>::visit_expr
  10: <rustc_typeck::check::writeback::WritebackCx as rustc_hir::intravisit::Visitor>::visit_expr
  11: <rustc_typeck::check::writeback::WritebackCx as rustc_hir::intravisit::Visitor>::visit_expr
  12: <rustc_typeck::check::fn_ctxt::FnCtxt>::resolve_type_vars_in_body
  13: <rustc_infer::infer::InferCtxtBuilder>::enter::<&rustc_middle::ty::context::TypeckResults, <rustc_typeck::check::inherited::InheritedBuilder>::enter<rustc_typeck::check::typeck_with_fallback<rustc_typeck::check::typeck::{closure#0}>::{closure#1}, &rustc_middle::ty::context::TypeckResults>::{closure#0}>
  14: rustc_typeck::check::typeck
  15: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, &rustc_middle::ty::context::TypeckResults>>
  16: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck
  17: rustc_typeck::check::typeck
  18: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, &rustc_middle::ty::context::TypeckResults>>
  19: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck
  20: <rustc_middle::hir::map::Map>::par_body_owners::<rustc_typeck::check::typeck_item_bodies::{closure#0}>
  21: rustc_typeck::check::typeck_item_bodies
  22: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), ()>>
  23: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::typeck_item_bodies, rustc_query_impl::plumbing::QueryCtxt>
  24: <rustc_session::session::Session>::time::<(), rustc_typeck::check_crate::{closure#7}>
  25: rustc_typeck::check_crate
  26: rustc_interface::passes::analysis
  27: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorReported>>>
  28: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
  29: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorReported>>
  30: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorReported>>
  31: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
  32: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.59.0-nightly (cfa3fe5af 2021-12-31) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `r`
#1 [typeck] type-checking `r::{closure#0}`
#2 [typeck_item_bodies] type-checking all item bodies
#3 [analysis] running analysis passes on this crate
end of query stack
Some errors have detailed explanations: E0308, E0425, E0559.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `playground` due to 7 previous errors
```

</p>
</details>

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"terrarier2111"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/src/test/ui/typeck/issue-92481.rs
+++ b/src/test/ui/typeck/issue-92481.rs
@@ -1,0 +1,14 @@
+//check-fail
+
+#![crate_type="lib"]
+
+fn r({) {
+    Ok {             //~ ERROR mismatched types [E0308]
+        d..||_=m
+    }
+}
+//~^^^^^ ERROR expected parameter name, found `{`
+//~| ERROR expected one of `,`, `:`, or `}`, found `..`
+//~^^^^^ ERROR cannot find value `d` in this scope [E0425]
+//~| ERROR cannot find value `m` in this scope [E0425]
+//~| ERROR variant `Result<_, _>::Ok` has no field named `d` [E0559]

--- a/src/test/ui/typeck/issue-92481.stderr
+++ b/src/test/ui/typeck/issue-92481.stderr
@@ -1,0 +1,62 @@
+error: expected parameter name, found `{`
+  --> $DIR/issue-92481.rs:5:6
+   |
+LL | fn r({) {
+   |      ^ expected parameter name
+
+error: expected one of `,`, `:`, or `}`, found `..`
+  --> $DIR/issue-92481.rs:5:6
+   |
+LL | fn r({) {
+   |      ^ unclosed delimiter
+LL |     Ok {
+LL |         d..||_=m
+   |          -^
+   |          |
+   |          help: `}` may belong here
+
+error[E0425]: cannot find value `d` in this scope
+  --> $DIR/issue-92481.rs:7:9
+   |
+LL |         d..||_=m
+   |         ^ not found in this scope
+
+error[E0425]: cannot find value `m` in this scope
+  --> $DIR/issue-92481.rs:7:16
+   |
+LL |         d..||_=m
+   |                ^ not found in this scope
+
+error[E0559]: variant `Result<_, _>::Ok` has no field named `d`
+  --> $DIR/issue-92481.rs:7:9
+   |
+LL |         d..||_=m
+   |         ^ field does not exist
+   |
+  ::: $SRC_DIR/core/src/result.rs:LL:COL
+   |
+LL |     Ok(#[stable(feature = "rust1", since = "1.0.0")] T),
+   |     -- `Result<_, _>::Ok` defined here
+   |
+help: `Result<_, _>::Ok` is a tuple variant, use the appropriate syntax
+   |
+LL |     Result<_, _>::Ok(/* fields */)
+   |
+
+error[E0308]: mismatched types
+  --> $DIR/issue-92481.rs:6:5
+   |
+LL |   fn r({) {
+   |           - help: a return type might be missing here: `-> _`
+LL | /     Ok {
+LL | |         d..||_=m
+LL | |     }
+   | |_____^ expected `()`, found enum `Result`
+   |
+   = note: expected unit type `()`
+                   found enum `Result<_, _>`
+
+error: aborting due to 6 previous errors
+
+Some errors have detailed explanations: E0308, E0425, E0559.
+For more information about an error, try `rustc --explain E0308`.

--- a/src/test/ui/typeck/issue-92481.stderr
+++ b/src/test/ui/typeck/issue-92481.stderr
@@ -32,11 +32,9 @@ error[E0559]: variant `Result<_, _>::Ok` has no field named `d`
    |
 LL |         d..||_=m
    |         ^ field does not exist
+  --> $SRC_DIR/core/src/result.rs:LL:COL
    |
-  ::: $SRC_DIR/core/src/result.rs:LL:COL
-   |
-LL |     Ok(#[stable(feature = "rust1", since = "1.0.0")] T),
-   |     -- `Result<_, _>::Ok` defined here
+   = note: `Result<_, _>::Ok` defined here
    |
 help: `Result<_, _>::Ok` is a tuple variant, use the appropriate syntax
    |

